### DISCUSSION TITLE
Ideas for annual undergraduate project

### DISCUSSION THREAD

**Discussion by erelsgl** [cvxpy/cvxpy#2176] [General]
Hi. Last year, a team of my students did a project that ended up [merged into `cvxpyanalyzer`](https://github.com/cvxpy/cvxpyanalyzer/pull/2). Now, I have a new team for the coming year. We are looking for a project that is more closely related to optimization, maybe based on a recent reseach paper, that could potentailly be merged into `cvxpy`. Do you have any ideas?

**Comment by SteveDiamond** [comment]
We discussed internally and one possibility is implementing dual variables for the ND power cone:
https://github.com/cvxpy/cvxpy/blob/e6c2a125e9e46273ef7c67e8136e00f9a94ce47b/cvxpy/constraints/power.py#L135

Another idea is to finish this PR adding``__rpow__`` support: https://github.com/cvxpy/cvxpy/issues/1700

But merging into cvxpy will require a lot more rounds of iteration than we did last time.

**Reply by SteveDiamond** [reply]
More ideas:
#1737 
#1529 
#1540

**Reply by erelsgl** [reply]
Which of these ideas is sufficiently substantial for an annual project of 3 students?

**Reply by SteveDiamond** [reply]
They may be too small for that.

**Comment by erelsgl** [comment]
OK, cool! Can you send references to papers that give background for the basic concepts?

**Reply by erelsgl** [reply]
Can you give a simple usage example for the ND power cone?

**Reply by rileyjmurray** [reply]
@erelsgl, one prominent application is in game theory, where the ND power cone is the natural representation for (weighted) Nash social welfare:
```math
\max_{x \in \Omega} \prod_i x_i^{\alpha_i} 
```
Right now this is usually handled by taking a log
```math
\max_{x \in \Omega} \sum_i \alpha_i \log(x_i)
```
This equivalent problem will have different numerical properties, which may affect solver performance (e.g., accuracy or time to solution) for the worse. So it would be nice if we didn't have to rely on this reformulation.

A more compelling example is in convex relaxations for nonconvex "generalized" polynomial optimization problems. I wrote a (rather technical) paper about this during my PhD work (see [Section 5.4 of my thesis](https://thesis.library.caltech.edu/14169/5/Riley%20Murray%20thesis%2C%20June%201%20deposit.pdf)). If you think your students might be up to it, then I can connect them with an undergrad I supervised at Caltech who worked on this for a summer.

Another possibility is improving how we use ND power cone constraints; see #1222.

**Reply by PTNobel** [reply]
Another paper that discusses the ND power cone (and includes examples) is here: https://arxiv.org/pdf/2305.12275.pdf

**Reply by erelsgl** [reply]
To be more concrete: right now, when I try to solve the problem 
```
cvxpy.Problem(objective=cvxpy.Maximize(x*y), constraints=[x+y==10, x>=0, y>=0])
```
I get an error: 
```
cvxpy.error.DCPError: Problem does not follow DCP rules. Specifically:
The objective is not DCP. Its following subexpressions are not:
var1[0] @ var1[1]
```

* The goal is that such problems could be solved directly (and yield the solution x=5, y=5). right?

* Besides problems such as maximizing a product or a weighted product, what are some other examples of `cvxpy` problems that generate an error now, and that should be solvable after the extension?

**Reply by erelsgl** [reply]
@rileyjmurray @PTNobel just a reminder about my question..

**Reply by SteveDiamond** [reply]
Someone implemented dual variable support for the NDPower constraint in cvxpy 1.4. Apologies for the poor coordination, we forgot you all were looking at this.

**Reply by erelsgl** [reply]
@SteveDiamond  It's OK, they still did not start working on it. So, what similar projects are available now?

**Reply by erelsgl** [reply]
@SteveDiamond is it possible to have a Zoom meeting to discuss potential projects?

**Reply by SteveDiamond** [reply]
Can you join the discord? https://discord.gg/4urRQeGBCr

**Comment by rileyjmurray** [comment]
We could also look at improving ND power cone support by actually using it
in atoms like geo_mean.

**Reply by erelsgl** [reply]
@rileyjmurray can we schedule a Zoom meeting, so that I can better understand the issues?

**Reply by erelsgl** [reply]
@rileyjmurray can we schedule a Zoom meeting, so that I can better understand the issues?

**Comment by rileyjmurray** [comment]
Hi Erel. I won't have time to meet until the end of this week at the
earliest. Please follow up with me on Thursday.

**Comment by erelsgl** [comment]
@rileyjmurray Are you available today (Thursday)?

**Comment by erelsgl** [comment]
Done
*--- Dr. Erel Segal-Halevi  (http://erelsgl.github.io
<http://erelsgl.github.io>) ד"ר אראל סגל הלוי*
